src/messages.cpp

```diff
@@ -265,170 +265,580 @@ Sender::sender_type() const
+  uint32_t generation{ 0 };
+  ReuseGuard reuse_guard{ 0, 0, 0, 0 };
+
+  TLS_SERIALIZABLE(sender, generation, reuse_guard)
+};
+
+struct MLSSenderDataAAD
+{
+  const bytes& group_id;
+  const epoch_t epoch;
+  const ContentType content_type;
+
+  TLS_SERIALIZABLE(group_id, epoch, content_type)
+};
+
+MLSCiphertext
+MLSCiphertext::protect(MLSMessageContentAuth content_auth,
+                       CipherSuite suite,
+                       const LeafIndex& index,
+                       GroupKeySource& keys,
+                       const bytes& sender_data_secret,
+                       size_t padding_size)
+{
+  // Pull keys from the secret tree
+  auto content_type = content_auth.content.content_type();
+  auto [generation, reuse_guard, content_keys] = keys.next(content_type, index);
+
+  // Encrypt the content
+  auto content_pt = marshal_ciphertext_content(
+    content_auth.content, content_auth.auth, padding_size);
+  auto content_aad = tls::marshal(MLSCiphertextContentAAD{
+    content_auth.content.group_id,
+    content_auth.content.epoch,
+    content_auth.content.content_type(),
+    content_auth.content.authenticated_data,
+  });
+
+  auto content_ct = suite.hpke().aead.seal(
+    content_keys.key, content_keys.nonce, content_aad, content_pt);
+
+  // Encrypt the sender data
+  auto sender_data_pt = tls::marshal(MLSSenderData{
+    var::get<LeafNodeRef>(content_auth.content.sender.sender),
+    generation,
+    reuse_guard,
+  });
+  auto sender_data_aad = tls::marshal(MLSSenderDataAAD{
+    content_auth.content.group_id,
+    content_auth.content.epoch,
+    content_auth.content.content_type(),
+  });
+
+  auto sender_data_keys =
+    KeyScheduleEpoch::sender_data_keys(suite, sender_data_secret, content_ct);
+
+  auto sender_data_ct = suite.hpke().aead.seal(sender_data_keys.key,
+                                               sender_data_keys.nonce,
+                                               sender_data_aad,
+                                               sender_data_pt);
+
+  return MLSCiphertext{
+    std::move(content_auth.content),
+    std::move(sender_data_ct),
+    std::move(content_ct),
+  };
+}
+
+std::optional<MLSMessageContentAuth>
+MLSCiphertext::unprotect(CipherSuite suite,
+                         const TreeKEMPublicKey& tree,
+                         GroupKeySource& keys,
+                         const bytes& sender_data_secret) const
+{
+  // Decrypt and parse the sender data
+  auto sender_data_keys =
+    KeyScheduleEpoch::sender_data_keys(suite, sender_data_secret, ciphertext);
+  auto sender_data_aad = tls::marshal(MLSSenderDataAAD{
+    group_id,
+    epoch,
+    content_type,
+  });
+
+  auto sender_data_pt = suite.hpke().aead.open(sender_data_keys.key,
+                                               sender_data_keys.nonce,
+                                               sender_data_aad,
+                                               encrypted_sender_data);
+  if (!sender_data_pt) {
+    return {};
```

Comment:
should the return be std::nullopt instead ?

Reply:
Sure.  It's equivalent, since `std::optional<T>` default-initializes to `std::nullopt`.  But it's clearer to be explicit.